common/helpers/src/avro.rs

```diff
@@ -0,0 +1,248 @@
+use crate::types::*;
+use apache_avro::{schema::Schema, types::Record, Codec, Reader, Writer};
+use std::{collections::HashMap, panic, str};
+
+/// Represents error types returned by the `avro` module.
+#[derive(thiserror::Error, Debug)]
+pub enum AvroError {
+	#[error("I/O error")]
+	Io(#[from] std::io::Error),
+	#[error("Avro error")]
+	Avro(#[from] apache_avro::Error),
+	#[error("Invalid avro schema: {0}")]
+	InvalidSchema(String),
+	#[error("Invalid avro records")]
+	InvalidRecords(),
+}
+
+/// Function to convert a raw schema into serialized Avro schema.
+/// If schema is malformed or invalid, returns an error.
+/// # Arguments
+/// * `raw_schema` - raw schema to be converted
+/// # Returns
+/// * `Result<(Schema, Vec<u8>), AvroError>` - structured and serialized Avro schema
+/// # Examples
+/// ```
+/// use avro::*;
+/// let raw_schema = r#"{"type": "record", "name": "User", "fields": [{"name": "name", "type": "string"}, {"name": "favorite_number", "type": "int"}]}"#;
+/// let schema_result = fingerprint_raw_schema(raw_schema);
+/// assert!(schema_result.is_ok());
+/// let serialized_schema = schema_result.unwrap().1;
+pub fn fingerprint_raw_schema(raw_schema: &str) -> Result<(Schema, Vec<u8>), AvroError> {
+	// parse_str will fail if schema is not valid
+	panic::catch_unwind(|| {
+		let schema_result = Schema::parse_str(raw_schema);
+		// match the result of the parse_str
+		match schema_result {
+			Ok(schema) => {
+				// get the fingerprint of the schema
+				let schema_canonical_form = schema.canonical_form();
+				// return the schema and the fingerprint
+				Ok((schema, schema_canonical_form.as_bytes().to_vec()))
+			},
+			Err(e) => Err(AvroError::Avro(e)),
+		}
+	})
+	.unwrap_or_else(|_| {
+		// if the unwind is caught, return an error
+		Err(AvroError::InvalidSchema(raw_schema.to_string()))
+	})
+}
+
```

Comment:
Given a valid raw avro schema translate it to avro which checks for validity on parsing or panic if schema is broken in either case return an error otherwise return a serialized schema to be stored on chain

Reply:
? Handles the panic and return error no need to panic_unwind